tests/tensor/test_extra_ops.py

```diff
@@ -291,47 +290,71 @@ def ref(data, w=None, minlength=None):
-        self.op = DiffOp()
-
     def test_diffOp(self):
         x = matrix("x")
         a = np.random.random((30, 50)).astype(config.floatX)
 
         f = aesara.function([x], diff(x))
         assert np.allclose(np.diff(a), f(a))
 
-        for axis in range(len(a.shape)):
-            for k in range(TestDiffOp.nb):
-                g = aesara.function([x], diff(x, n=k, axis=axis))
-                assert np.allclose(np.diff(a, n=k, axis=axis), g(a))
+        for axis in (-2, -1, 0, 1):
+            for n in (0, 1, 2, a.shape[0], a.shape[0] + 1):
+                g = aesara.function([x], diff(x, n=n, axis=axis))
+                assert np.allclose(np.diff(a, n=n, axis=axis), g(a))
+
+    @pytest.mark.parametrize(
+        "x_type",
+        (
+            at.TensorType("float64", (None, None)),
+            at.TensorType("float64", (None, 30)),
+            at.TensorType("float64", (10, None)),
+            at.TensorType("float64", (10, 30)),
+        ),
+    )
+    def test_output_type(self, x_type):
+        x = x_type("x")
+        x_test = np.empty((10, 30))
+        for axis in (-2, -1, 0, 1):
+            for n in (0, 1, 2, 10, 11):
+                out = diff(x, n=n, axis=axis)
+                out_test = np.diff(x_test, n=n, axis=axis)
+                for i in range(2):
+                    if x.type.shape[i] is None:
+                        assert out.type.shape[i] is None
+                    else:
+                        assert out.type.shape[i] == out_test.shape[i]
 
     def test_infer_shape(self):
         x = matrix("x")
         a = np.random.random((30, 50)).astype(config.floatX)
 
-        self._compile_and_check([x], [self.op(x)], [a], self.op_class)
+        # Test default n and axis
+        self._compile_and_check([x], [DiffOp()(x)], [a], DiffOp)
 
-        for axis in range(len(a.shape)):
-            for k in range(TestDiffOp.nb):
-                self._compile_and_check(
-                    [x], [diff(x, n=k, axis=axis)], [a], self.op_class
-                )
+        for axis in (-2, -1, 0, 1):
+            for n in (0, 1, 2, a.shape[0], a.shape[0] + 1):
+                self._compile_and_check([x], [diff(x, n=n, axis=axis)], [a], DiffOp)
 
     def test_grad(self):
-        x = vector("x")
         a = np.random.random(50).astype(config.floatX)
 
-        aesara.function([x], grad(at_sum(diff(x)), x))
-        utt.verify_grad(self.op, [a])
+        # Test default n and axis
+        utt.verify_grad(DiffOp(), [a])
+
+        for n in (0, 1, 2, a.shape[0]):
+            utt.verify_grad(DiffOp(n=n), [a], eps=7e-3)
+
+    def test_wrong_grad(self):
+        # Gradient is wrong when n is larger than the input size. Until it is fixed,
+        # this test ensures the behavior is documented
+        a = np.random.random(10).astype(config.floatX)
+        with pytest.raises(ValueError, match="^argument element 0 has wrong shapes"):
+            utt.verify_grad(DiffOp(n=11), [a], eps=7e-3)
```

Comment:
Fixing this would require changing the gradient logic or adding a bunch of `ifelse` in the gradient expression. Not sure if worth it